### ISSUE TITLE
Sizes on screen, LH and RH rectangles, and iPad and iPhone

### ISSUE THREAD

**Issue by gdcumming** [gfmoore/ESCI-JS#4]
Left hand rectangle looks about right for size, at 100% zoom. It’s easy to adjust the zoom level to suit for different screens, if desired. Goodie.

Resizing of RH rectangle is great. Overall, I suggest font size in RH rectangle (numbers on axes, label ‘mu’ when mu line is displayed) should be a little larger. 

On iPad in landscape, only about 90% of left hand rectangle is visible (down to about middle of panel 7). RH rectangle adjusts well. The speed control makes surprisingly little difference, fastest being only about double the slowest speed, and only about 2-3 per second. Sometimes we seem to get a display update only after every second sample. Sometimes stopping a run, clearing, then starting another run gives, straight away, a first display of 5 or 10 or 20 or something samples—the dance is already half way down the screen. This seems to happen only when speed is at or close to max. When it happens, the count of ‘Samples taken’ seems to be correct.

On iPad it doesn’t seem to be possible to slightly change zoom (careful squeezing of two fingers) to get all of LH rectangle on screen. It is possible, by very careful slow swipe up, to shift whole display and see all (or almost all) of LH rectangle. Overall, it looks great on iPad, very usable, the partial visibility of LH rectangle being the only real issue.

On iPhone in landscape (my 6S has the old small screen size) the RH rectangle adjusts well, the LH rectangle visible only down to about middle of panel 5. The speed control gives a very good and wide range of speeds, unlike on iPad. With great care it’s possible to scroll display higher to see lower parts of LH rectangle. Such scrolling often seems to clear and restart the sampling. For a tiny screen, it all works surprisingly well, and looks great. Need to take great care, with a very thin finger, to touch the controls accurately. The drop downs for choosing N and C work by expanding and showing the value options. User scrolls to choose, then returns to the main display to find this has been expanded, so need to use two fingers to zoom to desired display size. 

Maybe we don’t worry too much about tiny screens, unless and until we can have an app, maybe way in the future.

**Comment by gfmoore** [comment]
Uhmm.  I don't think I know yet how to handle this. 
It is a problem with UI design trying to do so much on tinier screens. I have an android tablet and phone so I'll have a look and see what I can do. 
I'm not sure that I can develop an app as such, it's a huge job and after a go at Android development I decided it was too much trouble. I can do most things with web pages. It just needs two or three versions for each type of device and that is probably the proper way to go.

You might want to have a think about the design of the UI for a tablet in landscape, and vertical and a phone likewise. 

My initial thoughts are that we don't really want 5 different UIs, possibly narrow it down to 3.

Laptop/tv/projector landscape (tablet/phone) portrait (tablet/phone)

On the tablet. phone you would probably need to swipe to get at the settings and then scroll down the different panels. Controls on the UI should be much bigger. Also introduces the tooltips issue, but with bigger controls there is more space to the side perhaps for a touchable area for help text.

Lots of work.

I'll see what I can do with the tablet in landscape anyway, I might be able to add a scroll to left hand control panel, or a swipe or something...

(I'm not giving up...)

**Comment by gdcumming** [comment]
I suspect always-landscape may be a good approach.

The app idea (esp. for phone) really is an outrageous suggestion. Definitely bottom of the list.

Maybe, at most, consider whether small tweaks to the basic large-screen software could make it run well (i.e. even better) on tablet. Achieving good sizing of the LH and RH rectangles may be the main thing. Plus coping with touch rather than mouse.

**Comment by gfmoore** [comment]
Right I am going to concentrate on getting a tablet sorted now. I know you like them :)

BUT not till next weekend. I need a break. :)

**Comment by gfmoore** [comment]
Can you have a look and see of this is any better, added a vertical scroll on lhs.

**Comment by gdcumming** [comment]
0.3.28

iPhone (Safari)
This can run so fast!
Scrolling of LH panel works well--nice!

iPad (Safari)
This also is faster--great.
I can't get scrolling of LH panel (separately from RH panel) to work at all.
Also layout in RH panel gets itself messed up: pop curve drops down relative to its x axis and data point circles. Bottom axis moves up. Not totally consistent. I'll try to send a screen shot.

**Comment by gfmoore** [comment]
I don't know why lh panel won't scroll on ipad. I
The problem is that it is impossible to detect what sort of device is using the app. I know that doesn't make sense, but it is an open topic on StackOverflow for the last ten years!!!

I optioned that to detect if a device had a touch screen (including your surface device no doubt) would get a scroll bar.
However, I don't know if the ipad is responding to the touchable test. I'll post a version on my  NAS box and you can tell me if you get an alert.
http://212.159.76.205/ESCI-JS/  

I could make the left panel always scrollable, but it not a pretty default scroll bar on a laptop. (There are others, but then introducing more code and opportunities for things to break :sigh)

I'm also not sure what I did to make it run faster? Though I have done lot's of things. It's acceptable speed on my old Samsung tablet. The population fill is pretty slow. 

I'm looking into the population curve not fitting into the right hand panel top section - This probably broke when I made the curves show the correct proportions when sigma changed. See what I mean about fix one issue introduce another. It's very hard to be all things. (Apparently professional app designers have different urls for different devices, so you have a mobile version and a desktop version, but nowadays the lines between them are so blurred I don't think that makes sense)

Yes the right hand panel was scaled so that the top 25% is for the curve, the bottom 75% for the dropping means. Again this is the issue with web based design approach being able to flow and fit as opposed to the old absolute positioning approach.
When we just had vga it was fine, but nowadays with so many aspect ratios and densities it's a nightmare.

**Comment by gdcumming** [comment]
0.3.30

iPhone still nice, also fast.

iPad now slow, and still no scroll.

Not so good, as you say, to see H and V scroll bars for LH rectangle on the computer big screen. So, perhaps, for this feature, let's step back a version or two, at least far enough to remove these?

Bob and I zoomed this morning (more later) but agreed that we suggest to you that getting things good on laptop and desktop should be the priority. If that means things work, or nearly work well, on phones and tablets, then that's a (big) bonus, but not for the moment core business.

In other words, let's put an app version, and/or tailoring for phones and tablets, on the 'Maybe, sometime' list? OK with you?

We keenly appreciate the challenges of diverse devices, diverse screen characteristics, etc.

On the iPad I went to that NAS url and did get a message 'touchable', then saw version 0.3.29. Slow, and pop curve was displayed as extending down below H axis line, overlapping the display of the data values for last sample.

**Comment by gfmoore** [comment]
I wonder what it is about ipad that causes issues? Can't understand why no scroll bar? What version of IOS is it on, I'll do a search see if I can find anything on that version. Is it an older ipad? What memory does it have?
When you say slow then is there a difference if you don't fill the population curve? 

The population fill is probably a huge memory hog as there are so many it displays and all need to be in the DOM. 

But yes put on the back burner for now. It's a pain testing tablets anyway as they don't have a debug (developer) console for the browser, so you need to do debugging the old way - lots of alerts or writing to the screen. Plus there's no direct link between the IDE (VS Code) and the tablet.

I will keep looking at my tablet and phone though, so please feel free to keep posting bugs or images here. 

I'll close it, but you can keep adding to it and then we can bring it back to open if we want to.
